### PR TITLE
make waveform play marker position adjustable

### PR DESCRIPTION
**Pull request by Be-ing**
This supercedes #1520, which stalled because NTMusic deleted their GitHub account.

### CODE REVIEW

File: src/waveform/renderers/waveformrendererpreroll.h [hunk 1]
```diff
@@ -17,7 +17,7 @@ class WaveformRendererPreroll : public WaveformRendererAbstract {
 
   private:
     QColor m_color;
-
+
```
Comment (daschuer) [inline src/waveform/renderers/waveformrendererpreroll.h]: stray change

File: src/waveform/renderers/waveformwidgetrenderer.cpp [hunk 2]
```diff
@@ -12,6 +12,7 @@
 const int WaveformWidgetRenderer::s_waveformMinZoom = 1;
 const int WaveformWidgetRenderer::s_waveformMaxZoom = 10;
 const int WaveformWidgetRenderer::s_waveformDefaultZoom = 3;
+const double WaveformWidgetRenderer::s_DefaultPlayMarkerPosition = 0.5;
```
Comment (daschuer) [inline src/waveform/renderers/waveformwidgetrenderer.cpp]: s_default...

File: src/waveform/renderers/waveformwidgetrenderer.cpp [hunk 3]
```diff
@@ -142,29 +146,39 @@ void WaveformWidgetRenderer::onPreRender(VSyncThread* vsyncThread) {
...
         m_playPosVSample = m_playPos * m_trackPixelCount * m_visualSamplePerPixel;
 
-        m_firstDisplayedPosition = m_playPos - displayedLengthHalf;
-        m_lastDisplayedPosition = m_playPos + displayedLengthHalf;
+        double leftOffset = m_playMarkerPosition;
+        double rightOffset = 1.0 - m_playMarkerPosition;
+
+        double displayedLengthLeft = (static_cast<double>(getLength()) / m_trackPixelCount) * leftOffset;
+        double displayedLengthRight = (static_cast<double>(getLength()) / m_trackPixelCount) * rightOffset;
+/*
```
Comment (daschuer) [inline src/waveform/renderers/waveformwidgetrenderer.cpp]: use // for comments.

Comment (daschuer) [reply src/waveform/renderers/waveformwidgetrenderer.cpp]: or #ifdef

File: src/waveform/renderers/waveformwidgetrenderer.cpp [hunk 4]
```diff
@@ -142,29 +146,39 @@ void WaveformWidgetRenderer::onPreRender(VSyncThread* vsyncThread) {
...
+*/
+        m_firstDisplayedPosition = m_playPos - displayedLengthLeft;
+        m_lastDisplayedPosition = m_playPos + displayedLengthRight;
     } else {
         m_playPos = -1; // disable renderers
     }
 
     /*
     qDebug() << "m_group" << m_group
-             << "m_trackSamples" << m_trackSamples
```
Comment (daschuer) [inline src/waveform/renderers/waveformwidgetrenderer.cpp]: the same here.

File: src/waveform/renderers/waveformwidgetrenderer.h [hunk 5]
```diff
@@ -93,6 +94,14 @@ class WaveformWidgetRenderer {
 
     void setTrack(TrackPointer track);
 
+    double getPlayMarkerPosition() {
+        return m_playMarkerPosition;
+    }
+
+    void setPlayMarkerPosition(double newPos) {
+        m_playMarkerPosition = math_clamp(newPos, 0.0, 1.0);
```
Comment (daschuer) [inline src/waveform/renderers/waveformwidgetrenderer.h]: Is this actually an debug assert?

File: src/waveform/waveformwidgetfactory.cpp [hunk 6]
```diff
@@ -71,7 +71,8 @@ WaveformWidgetFactory::WaveformWidgetFactory() :
         m_vsyncThread(NULL),
         m_frameCnt(0),
         m_actualFrameRate(0),
-        m_vSyncType(0) {
+        m_vSyncType(0),
+        m_playMarkerPosition(0.5) {
```
Comment (daschuer) [inline src/waveform/waveformwidgetfactory.cpp]: can we use here: WaveformWidgetRenderer::s_DefaultPlayMarkerPosition; 
The constant should be probably moved.

File: src/waveform/waveformwidgetfactory.cpp [hunk 7]
```diff
@@ -252,6 +253,12 @@ bool WaveformWidgetFactory::setConfig(UserSettingsPointer config) {
         m_config->set(ConfigKey("[Waveform]","OverviewNormalized"), ConfigValue(m_overviewNormalized));
     }
 
+    int playMarkerPosition = m_config->getValueString(ConfigKey("[Waveform]","PlayMarkerPosition")).toInt(&ok);
+    if (ok) {
```
Comment (daschuer) [inline src/waveform/waveformwidgetfactory.cpp]: I do not understand the code here. Some comments can help.

File: src/waveform/waveformwidgetfactory.cpp [hunk 8]
```diff
@@ -252,6 +253,12 @@ bool WaveformWidgetFactory::setConfig(UserSettingsPointer config) {
         m_config->set(ConfigKey("[Waveform]","OverviewNormalized"), ConfigValue(m_overviewNormalized));
     }
 
+    int playMarkerPosition = m_config->getValueString(ConfigKey("[Waveform]","PlayMarkerPosition")).toInt(&ok);
```
Comment (daschuer) [inline src/waveform/waveformwidgetfactory.cpp]: Why int? 
double playMarkerPosition = m_config->getValue(ConfigKey("[Waveform]","PlayMarkerPosition"), 0.5);

File: src/waveform/waveformwidgetfactory.cpp [hunk 9]
```diff
@@ -466,6 +476,23 @@ void WaveformWidgetFactory::setOverviewNormalized(bool normalize) {
     }
 }
 
+void WaveformWidgetFactory::setPlayMarkerPosition(double position) {
+    //qDebug() << "setPlayMarkerPosition, position=" << position;
+    m_playMarkerPosition = position;
+    if (m_config) {
+        m_config->set(ConfigKey("[Waveform]", "PlayMarkerPosition"), ConfigValue(m_playMarkerPosition));
```
Comment (daschuer) [inline src/waveform/waveformwidgetfactory.cpp]: m_config->setValue(ConfigKey("[Waveform]", "PlayMarkerPosition"), m_playMarkerPosition);

File: src/waveform/waveformwidgetfactory.cpp [hunk 10]
```diff
@@ -466,6 +476,23 @@ void WaveformWidgetFactory::setOverviewNormalized(bool normalize) {
...
+    }
+
+    if (m_waveformWidgetHolders.size() == 0) {
```
Comment (daschuer) [inline src/waveform/waveformwidgetfactory.cpp]: this is redundant

File: src/waveform/waveformwidgetfactory.cpp [hunk 11]
```diff
@@ -466,6 +476,23 @@ void WaveformWidgetFactory::setOverviewNormalized(bool normalize) {
...
+        return;
+    }
+
+    for (int i = 0; i < m_waveformWidgetHolders.size(); i++) {
```
Comment (daschuer) [inline src/waveform/waveformwidgetfactory.cpp]: range based loop?

**Review by daschuer — CHANGES_REQUESTED** [review]
Works nice

File: src/waveform/renderers/waveformwidgetrenderer.h (hunk 5 above)
Comment (Be-ing) [reply src/waveform/renderers/waveformwidgetrenderer.h]: Oops, there is a [DEBUG_ASSERT in math_clamp](https://github.com/mixxxdj/mixxx/blob/master/src/util/math.h#L36), but it's not checking what I thought it was.

**Comment by Be-ing** [comment]
Ready for merge?

**Comment by daschuer** [comment]
Yes, Thank you. LGTM

**Pull request merged**